java/src/jmri/configurexml/DirtyManager.java

```diff
@@ -0,0 +1,149 @@
+
+    final private PropertyChangeListener _mgrListen = new ManagerListener();
+    final private PropertyChangeListener _beanListen = new BeanListener();
+
+    public DirtyManager() {
+    }
+
+    /**
+     * When a manager notifies ConfigXmlManager that it is responsible for creating a portion of
+     * the PanelPro data file, the manager is forwarded here so that its changes can be captured.
+     * @param o The manager object.
+     */
+    protected void addManager(Object o) {
+        log.info("++ 1 dm add manager :: {}", o.getClass().getName());
+        var cname = o.getClass().getName();
+        if (cname.contains("DefaultConditionalManager")) return;  // Special handling required.
+        if (o instanceof jmri.managers.AbstractManager) {
+            Manager<?> m = (Manager<?>) o;
+            m.addPropertyChangeListener(_mgrListen);
+            loadBeans(o);
+        } else {
+            log.info("!! Unable to identify class:  {}", o);
+        }
+    }
+
+    /**
+     * Attach a listener to all of the children of a manager to listen for bean changes.
+     * @param o The bean manager object.
+     */
+    private void loadBeans(Object o) {
+        Manager<?> m = (Manager<?>) o;
+        m.getNamedBeanSet().forEach((bean) -> {
+            bean.removePropertyChangeListener(_beanListen);
+            bean.addPropertyChangeListener(_beanListen);
+        });
+    }
+
+    /**
+     * Set the dirty flag true or false.
+     * True occurs when changes are reported by listener events and other special cases.
+     * After a Store, the flag is set to false.
+     * @param dirty Either true or false.
+     */
+    public void setDirty(boolean dirty) {
+        log.info("## dirty flag = {}, loading = {}", dirty, _loading);
+        if (!_loading) {
+            _dirty = dirty;
+        }
+    }
+
+    /**
+     * This is used by storeIfNeeded which is invoked by the shutdown manager.  If true, provide an
+     * option to do the Store process before the shutdown occurs.
+     * @return the dirty state.
+     */
+    public boolean isDirty() {
+        return _dirty;
+    }
+
+    /**
+     * This is used to ignore the changes that naturally occur data loading.
+     * Set by the PanelPro data load processes to before loading and false when done.
+     * @param loading Either true or false.
+     */
+    public void setLoading(boolean loading) {
+        log.info("@@ loading flag = {}", loading);
+        _loading = loading;
+    }
+
+    /**
+     * This in invoked by the various dirty processes to see if a load is in process.
+     * @return the loading state.
+     */
+    public boolean isLoading() {
+        return _loading;
+    }
+
+    /**
+     * If the dirty state is true, provide the user with the opporutnity to do a Store before
+     * the shutdown completes.  This is invoked by the shutdown manager.
+     */
+    public void storeIfNeeded() {
+        if (isDirty()) {
+            DirtyManagerDialog.showDialog();
```

Comment:
It would be good if this was:

        if (isDirty() && !GraphicsEnvironment.isHeadless()) {
            DirtyManagerDialog.showDialog();